libfreefare/mifare_desfire_key.c

```diff
@@ -124,16 +127,21 @@ mifare_desfire_key_get_version(MifareDESFireKey key)
 void
 mifare_desfire_key_set_version(MifareDESFireKey key, uint8_t version)
 {
+    if (key->type == T_AES) {
+	key->aes_version = version;
+	return;
+    }
+
     for (int n = 0; n < 8; n++) {
 	uint8_t version_bit = ((version & (1 << (7 - n))) >> (7 - n));
 	key->data[n] &= 0xfe;
 	key->data[n] |= version_bit;
 	if (key->type == T_DES) {
 	    key->data[n + 8] = key->data[n];
-	} else {
+	} else if (key->type == T_3DES) {
 	    // Write ~version to avoid turning a 3DES key into a DES key
 	    key->data[n + 8] &= 0xfe;
-	    key->data[n + 8] |= ~version_bit;
+	    key->data[n + 8] |= !version_bit;
```

Comment:
D'oh :scream: !

So, yes, `~0x00` → `0xFF` and `~0x01` → `0xFE` so all key data is corrupt, this is so obvious I failed to see it…

That being said, or'ing with the result of a boolean expression looks odd to me: I am not 100% sure that C states that the result of a boolean expression is always 0 or 1, and if it's not guaranteed, it would be "safer" to:

```c
key->data[n + 8] |= (version_bit) ? 0x00 : 0x01;
```

I failed to find something that tells me that any boolean expression shall return 0 or 1.  If you have some information in respect to this, please educate me and ignore the preceding :smile:

Reply:
IEEE Std 1003.1-2001 defines a boolean `true` to be 1 and `false` to be zero. So, yes, I'm 100% sure this isn't relying on undefined behavior.

Reply:
Great!  I have to admit that I feel like the "long" version is easier to read than the '!' one, but I can live with it :+1: